src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,206 @@
+/-
+Copyright (c) 2020 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Joseph Myers.
+-/
+import algebra.add_torsor
+import linear_algebra.basis
+
+noncomputable theory
+
+/-!
+# Affine spaces
+
+This file defines affine spaces and subspaces and affine maps and the
+affine span of a set of points.
+
+## Implementation notes
+
+This file is very minimal and many things are surely omitted. Most
+results can be deduced from corresponding results for vector spaces.
+The variables `k` and `V` are explicit rather than implicit arguments
+to lemmas because otherwise the elaborator sometimes has problems
+inferring appropriate types and type class instances.  Definitions of
+affine spaces vary as to whether a space with no points is permitted;
+here, we require a nonempty type of points (via the definition of
+torsors requiring a nonempty type).
+
+## References
+
+* https://en.wikipedia.org/wiki/Affine_space
+* https://en.wikipedia.org/wiki/Principal_homogeneous_space
+
+-/
+
+/-- `affine_space` is an abbreviation for `add_torsor` in the
+case where the group is a vector space. -/
+@[nolint unused_arguments]
+abbreviation affine_space (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V]
```

Comment:
The only instance that's really needed here is `[add_group V]`, which means every time this declaration is used, Lean will run two unnecessary type class searches. Is there a real reason to require them in this definition? Presumably they're needed in declarations below, and of course it's fine to require them there.

Reply:
Everything that only requires an `add_group` should already be in `algebra.add_torsor`. The definitions in this file depend on `submodule.span` or `linear_map`; some of them might work for modules in place of vector spaces, but not more generally than that.

Reply:
(Thus, this definition simply corresponds to one of the conventional definitions of what an affine space is.)

Reply:
I think the essential question here is: is it appropriate to have the `abbreviation affine_space` at all (because if so, it should have the arguments appropriate to the concept of an affine space), or should this file just comment on what an affine space is without defining `affine_space`, and then define other things such as `affine_subspace` and `affine_map` in terms of `add_torsor` together with whatever the requirements on `V` are for each individual definition?

Reply:
I think it's ok to have the definition. You could nevertheless "cheat" and drop the `field` assumption when it's not needed, just to make life easier for lean. Most theorems and lemmas will still have the assumption.

Reply:
In fact all the definitions and proofs here work for a `module` over a `ring` in place of a vector space (although there may well be some in future that don't), and it seems that affine spaces over a module are a real mathematical concept (although not mentioned on Wikipedia). One proof fails with a `semimodule` over a `semiring` (using current mathlib master); I think the result `vsub_mem_vector_span_of_mem_span_points_of_mem_span_points` is still true in that case, but would need a different, more complicated proof as `neg_one_smul` is no longer valid in the `semimodule` case (and I'm not sure affine spaces over semimodules are actually useful objects).

Reply:
I agree that it's best to keep the definition, but without `[field k]`. You can make the doc-string more explicit about the fact that for all uses of the definition, we will need `[field k]`. It would be fine to be even more explicit, saying "but we omit the typeclass `[field k]` in the type synonym itself to simplify typeclass search".

Reply:
I've removed `[field k]` and thus `[vector_space k V]` from the `affine_space` definition, expanding the comment as suggested. It still needs `@[nolint unused_arguments]` unless we also remove `(k : Type*)` as an argument, should that be removed as well?

Reply:
No, `k` itself should stay. It's pretty common for type synonyms to need a `@[nolint unused_arguments]`.